ext/dom/xml_document.c

```diff
@@ -0,0 +1,204 @@
+
+/* Living spec never creates explicit namespace declaration nodes.
+ * They are only written upon serialization but never appear in the tree.
+ * So in principle we could just ignore them outright.
+ * However, step 10 in https://html.spec.whatwg.org/multipage/parsing.html#create-an-element-for-the-token
+ * requires us to have the declaration as an attribute available */
+static void dom_mark_namespaces_as_attributes_too(xmlDocPtr doc)
+{
+	xmlNodePtr node = doc->children;
+	while (node != NULL) {
+		if (node->type == XML_ELEMENT_NODE) {
+			dom_ns_compat_mark_attribute_list(node->nsDef);
+
+			if (node->children) {
+				node = node->children;
+				continue;
+			}
+		}
+
+		if (node->next) {
+			node = node->next;
+		} else {
+			/* Go upwards, until we find a parent node with a next sibling, or until we hit the base. */
+			do {
+				node = node->parent;
+				if (node == NULL) {
+					return;
+				}
+			} while (node->next == NULL);
+			node = node->next;
+		}
+	}
+}
+
+void dom_mark_namespaces_for_copy_based_on_copy(xmlNodePtr copy, const xmlNode *original)
+{
+	xmlNodePtr copy_current = copy;
+	const xmlNode *original_current = original;
+	while (copy_current != NULL) {
+		ZEND_ASSERT(original_current != NULL);
+
+		if (copy_current->type == XML_ELEMENT_NODE) {
+			dom_ns_compat_copy_attribute_list_mark(copy_current->nsDef, original_current->nsDef);
+
+			if (copy_current->children) {
+				copy_current = copy_current->children;
+				original_current = original_current->children;
+				continue;
+			}
+		}
+
+		if (copy_current->next) {
+			copy_current = copy_current->next;
+			original_current = original_current->next;
+		} else {
+			/* Go upwards, until we find a parent node with a next sibling, or until we hit the base. */
+			do {
+				copy_current = copy_current->parent;
+				if (copy_current == NULL) {
+					return;
+				}
+				original_current = original_current->parent;
+			} while (copy_current->next == NULL);
+			copy_current = copy_current->next;
+			original_current = original_current->next;
+		}
+	}
+}
+
+PHP_METHOD(DOM_XMLDocument, createEmpty)
+{
+	const char *version = NULL;
+	size_t encoding_len = strlen("UTF-8");
+	const char *encoding = "UTF-8";
+	size_t version_len;
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|ss", &version, &version_len, &encoding, &encoding_len) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	xmlCharEncodingHandlerPtr handler = xmlFindCharEncodingHandler(encoding);
+
+	if (handler != NULL) {
+		xmlCharEncCloseFunc(handler);
+	} else {
+		zend_argument_value_error(2, "is not a valid document encoding");
+		RETURN_THROWS();
+	}
+	
+    xmlDocPtr lxml_doc = xmlNewDoc((const xmlChar *) version);
+	if (UNEXPECTED(lxml_doc == NULL)) {
+		goto oom;
+	}
+
+	lxml_doc->encoding = xmlStrdup((const xmlChar *) encoding);
+
+	dom_object *intern = php_dom_instantiate_object_helper(return_value, dom_xml_document_class_entry, (xmlNodePtr) lxml_doc, NULL);
+	intern->document->is_modern_api_class = true;
+	return;
+
+oom:
```

Comment:
Jump labels should be indented by a single space to improve naive diff algorithms from misdetecting this as a function name for the hunk title.

Reply:
Wow TIL, that's annoying behaviour